onnx/defs/nn/defs.cc

```diff
@@ -1718,10 +1749,102 @@ ONNX_OPERATOR_SET_SCHEMA(
             "T",
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "T1",
+            {"tensor(bool)"},
+            "Constrain input training_mode to boolean tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           propagateShapeAndTypeFromFirstInput(ctx);
-          // TODO in training mode, it may be possible to infer some of
-          // the other outputs as well.
+          if (hasInputShape(ctx, 0)) {
+            propagateShapeFromInputToOutput(ctx, 0, 0);
+            auto& x_input_shape = getInputShape(ctx, 0);
+            int num_channels = 1;
+            if (static_cast<int>(x_input_shape.dim_size()) > 1 &&
+                x_input_shape.dim(1).has_dim_value()) {
+              num_channels = static_cast<int>(x_input_shape.dim(1).dim_value());
+            }
+
+            if (hasInputShape(ctx, 1)) {
+              auto& scale_input_shape = getInputShape(ctx, 1);
+              if (static_cast<int>(scale_input_shape.dim_size()) != 1 ||
+                  !scale_input_shape.dim(0).has_dim_value() ||
+                  static_cast<int>(scale_input_shape.dim(0).dim_value()) !=
```

Comment:
Why throw if it doesn't have a dim_value ? What if it has a dim_param ? Should we check to see if it matches the dim_param of the channel index in the input shape (if that is a dim_param too) ?